STM32 UART DRIVER :  add support for buffer circular mode

djiatsaf-st opened the pull request:
This PR introduces the implementation of the circular mode in the uart communication on the stm32 driver.

It is inspired by this commit  https://github.com/zephyrproject-rtos/zephyr/commit/72752802f18a70542a8fe79285a4a6419f5d5945#diff-54bdff4ff3a16e194201b66e24cada279fdfdd40dfd66b1df3df3a99d0f65cd0R1007 made by @fouge 

The cyclic mode will be configured from the devicetree and will also have the management of the dma buffer with the integration of the half full and full irq .

This implementation is compatible with all STM32 series that have the CIRC bit at the 5th position of the DMA_CCRx register.

A sample driver has been made available and tested on the C0, F0 and G0 series with communication speeds reaching up to 500kbps.

Fixes #79831

erian747 commented:
I have not analyzed the PR in detail, but saw some fragments of code that could possibly touching a solution to discussion in:
https://github.com/zephyrproject-rtos/zephyr/discussions/81104 which is about calling event callback in case of rx timeout (even with 0 bytes received)

djiatsaf-st committed:
include: zephyr: drivers: dma: add support for circular DMA

This macro will help us to enable the 5th bit CIRC on the DMA_CCRx
register for all stm32 series having this configuration.
Thus the DMA will be able to handle the circular buffers.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>
co-authored-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

djiatsaf-st committed:
drivers: serial: add circular mode in uart driver

the following implementations are added:
- set cyclic variable to handle circular/normal mode :
  enable dest/source_reload_en variable to set CIRC bit to 1 for RX or TX.
- DMA_STATUS_COMPLETE(0), DMA_STATUS_BLOCK(1) macros to handle half
 and full irq
- add ASYNC_UART_STATUS_TIMEOUT(3)  macro to work with  default mode
- add status input in uart_stm32_dma_rx_flush function  to handle
 async events depending on the  mode we are in.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>
co-authored-by: Cyril Fougeray <cyril.fougeray@worldcoin.org>

djiatsaf-st commented on samples/drivers/uart/ring_buffer/prj.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_RING_BUFFER=y
+CONFIG_LOG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
+CONFIG_PRINTK=y
```
Maybe I don't understand you well, you're talking about the config logs or something else? 

or you want me to put them in a config overlay? 

I have already removed the boards specific ones and it works with these above configs enabled.

djiatsaf-st replied:
> > Sample is now in samples: boards: st: uart
> 
> I don't think that's necessary.

Just for information, the change of location is not related to your suggestions mentioned above

djiatsaf-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -1117,11 +1125,16 @@ static inline void async_evt_rx_rdy(struct uart_stm32_data *data)
-	if (event.data.rx.len > 0) {
+		/* send event only for new data */
+		if (event.data.rx.len > 0) {
+			async_user_callback(data, &event);
```
I didn't do it just to preserve how normal mode was implemented in the previous version. 

For me it's better to leave it as is to avoid potential conflicts.

djiatsaf-st committed:
samples: boards: st: uart: add sample for uart circular mode

This code example  will demonstrate how to use uart
circular mode on STM32 boards.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

gautierg-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -1204,20 +1217,42 @@ static inline void async_timer_start(struct k_work_delayable *work,
+				   data->dma_rx.dma_channel, &stat) == 0) {
+			rx_rcv_len = data->dma_rx.buffer_length -
+				     stat.pending_length;
+		}
```
You're missing a `break;` here

djiatsaf-st replied:
Default is the last case, nothing to do after. It is possible to use it here but I think it is not necessary.

gautierg-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -1204,20 +1217,42 @@ static inline void async_timer_start(struct k_work_delayable *work,
+		data->dma_rx.offset = data->dma_rx.buffer_length / 2;
+
+		break;
+	default: /* likely ASYNC_UART_STATUS_TIMEOUT */
```
If it's likely, why not use `case ASYNC_UART_STATUS_TIMEOUT:` directly ?

djiatsaf-st replied:
Just to be careful, since we can't confirm the full control of the driver events. This will allow us not to break what had been done by default. 
We just handle the cases 1 (half complete) 0 (for full complete).

erwango commented on samples/boards/st/uart/circular_dma/src/main.c:
I'm still puzzled by the fact we're not able to use a blocking mechanism here.

djiatsaf-st replied:
I found what to do to make it work with a flag. I update the code

erwango commented on samples/boards/st/uart/circular_dma/src/main.c:
If there is no other consequence, then let's move to this solution to get to the most simple version of the sample

djiatsaf-st replied:
There will be a small consequence, the user will have the impression of losing data while it will be displayed at the next interruption.
I think I will remove this comment, originally it was for those who wanted to know how the mechanism works behind or have a similar result of the circular buffer as in the stm32cube examples.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -99,6 +99,14 @@ uint32_t lpuartdiv_calc(const uint64_t clock_rate, const uint32_t baud_rate)
 
+#ifdef CONFIG_UART_ASYNC_API
+#define STM32_ASYNC_STATUS_TIMEOUT (DMA_STATUS_BLOCK + 1)
+BUILD_ASSERT(STM32_ASYNC_STATUS_TIMEOUT != DMA_STATUS_COMPLETE,
+	     "STM32_ASYNC_STATUS_TIMEOUT must be different from DMA_STATUS_COMPLETE");
+BUILD_ASSERT(STM32_ASYNC_STATUS_TIMEOUT != DMA_STATUS_BLOCK,
+	     "STM32_ASYNC_STATUS_TIMEOUT must be different from DMA_STATUS_BLOCK");
```
These checks are no more required when using `#define STM32_ASYNC_STATUS_TIMEOUT (DMA_STATUS_BLOCK + 1)`

erwango commented on samples/boards/st/uart/circular_dma/boards/nucleo_c031c6.overlay:
```diff
@@ -0,0 +1,21 @@
+ */
+
+&usart2 {
+	dmas = <&dmamux1 0 53 (STM32_DMA_PERIPH_TX )>, //| STM32_DMA_MODE_CYCLIC
```
Clean up comment

djiatsaf-st commented on samples/boards/st/uart/circular_dma/src/main.c:
```diff
@@ -0,0 +1,135 @@
+	uint8_t rx_data[RX_BUF_SIZE];
+	size_t len;
+
+	while (1) {
+
+		if (flag) {
+
```
I did a test with this suggestion, but I don't get the data anymore, it seems that with this thread context, it needs to work continuously and only process data when the flag is raised.

erwango commented on samples/boards/st/uart/circular_dma/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_SERIAL=y
+CONFIG_UART_ASYNC_API=y
+CONFIG_RING_BUFFER=y
+CONFIG_DCACHE=n
```
Not required

erwango reviewed (changes requested):
Otherwise LGTM

fouge commented:
will try that hopefully today!
